prime-router/src/main/kotlin/fhirengine/translation/hl7/FhirTransformer.kt

```diff
@@ -0,0 +1,225 @@
+        return bundle
+    }
+
+    /**
+     * Transform the [bundle] using the elements in the given [schema] using [context] starting at the
+     * [focusResource] in the bundle. Set [debug] to true to enable debug statements to the logs.
+     */
+    private fun transformWithSchema(
+        schema: FhirTransformSchema,
+        bundle: Bundle,
+        focusResource: Base,
+        context: CustomContext = CustomContext(bundle, focusResource),
+        debug: Boolean = false
+    ) {
+        val logLevel = if (debug) Level.INFO else Level.DEBUG
+        logger.log(logLevel, "Processing schema: ${schema.name} with ${schema.elements.size} elements")
+        // Add any schema level constants to the context
+        // We need to create a new context, so constants exist only within their specific schema tree
+        val schemaContext = CustomContext.addConstants(schema.constants, context)
+
+        schema.elements.forEach { element ->
+            transformBasedOnElement(element, bundle, focusResource, schemaContext, debug)
+        }
+    }
+
+    /**
+     * Transform the [bundle] using [element] and [context] starting at the
+     * [focusResource] in the bundle. Set [debug] to true to enable debug statements to the logs.
+     */
+    internal fun transformBasedOnElement(
+        element: FHIRTransformSchemaElement,
+        bundle: Bundle,
+        focusResource: Base,
+        context: CustomContext,
+        debug: Boolean = false
+    ) {
+        val logLevel = if (element.debug || debug) Level.INFO else Level.DEBUG
+        logger.trace("Started processing of element ${element.name}...")
+        // Add any element level constants to the context
+        val elementContext = CustomContext.addConstants(element.constants, context)
+        var debugMsg = "Processed element name: ${element.name}, required: ${element.required}, "
+
+        // First we need to resolve a resource value if available.
+        val focusResources = getFocusResources(element.resource, bundle, focusResource, elementContext)
+        if (focusResources.isEmpty() && element.required == true) {
+            // There are no sources to parse, but the element was required
+            throw RequiredElementException(element)
+        } else if (focusResources.isEmpty()) debugMsg += "resource: NONE"
+
+        focusResources.forEachIndexed { index, singleFocusResource ->
+            // The element context must now get the focus resource
+            elementContext.focusResource = singleFocusResource
+            if (canEvaluate(element, bundle, singleFocusResource, elementContext)) {
+                when {
+                    // If this is a schema then process it.
+                    element.schemaRef != null -> {
+                        // Schema references can have new index references
+                        val indexContext = if (element.resourceIndex.isNullOrBlank()) elementContext
+                        else CustomContext.addConstant(
```

Comment:
Should this chunk have code coverage?

Reply:
It should eventually, but that'll be more relevant to [the story for repeating elements](https://app.zenhub.com/workspaces/platform-6182b02547c1130010f459db/issues/gh/cdcgov/prime-reportstream/7248). Writing tests for that and verifying that things work is probably the main work for that story, so I'll save it for then, but I still adapted this piece of the code over as a starting point.